demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
```

Comment:
The idea of finite state machines is that in the rest of your code you generally don't have a list of all the valid states. Admittedly I haven't been particularly strict about that here. But instead of treating states as a fancy enumeration:

```python
if obj.state in (ReadyState, WaitingState):
   # do something
```

you instead just call methods on the state, and you don't care what the state actually is:

```python
if obj.state.is_healthy():
    # do something
```

Also, in terms of expressing things like "Error while applying configuration", you could do that by simply asking each state for a `state.status_message()`, `state.status_message_color()`, `state.full_status_message()`.